src/coq_rules.ml

```diff
@@ -90,11 +118,48 @@ let create_ccoq sctx ~dir =
   ; coqpp  = rr "coqpp"
   }
 
+(* compute include flags and mlpack rules *)
+let process_ml_deps ~scope ~stdlib_dir libs =
+
+  (* coqdep expects an mlpack file next to the sources otherwise it
+   * will omit the cmxs deps *)
+  let ml_pack_file lib =
+    let plugin_name = Option.value_exn (Result.ok_exn (Lib.main_module_name lib)) in
+    let mlpack_file = Module.Name.uncapitalize plugin_name ^ ".mlpack" in
+    Path.relative (Lib.src_dir lib) mlpack_file
+  in
+
+  (* Pair of include flags and paths to mlpack *)
+  let ml_iflags, mlpack =
+    let lib_db = Scope.libs scope in
+    let loc = Loc.none in
+    match Lib.DB.find_many ~loc lib_db
+            List.(concat_map ~f:Dune_file.Lib_dep.to_lib_names libs) with
+    | Ok libs ->
+      Util.include_flags ~stdlib_dir libs, List.map ~f:ml_pack_file libs
+    | Error exn ->
+      (* TODO: Proper error handling *)
+      raise exn
+  in
+
+  (* If the mlpack files don't exist, don't fail *)
+  let rec maybe_paths files =
+    match files with
+    | [] -> Build.return ()
+    | file :: files ->
+      (Build.(if_file_exists file ~then_:(path file) ~else_:(return ())) &&&
+       maybe_paths files) >>^ fst
```

Comment:
In `format_rules`, there's something similar already: `depend_on_existing_paths`. In `build_system.ml`, we also have something not too different in `dep_rec_internal`. Can we generalize this to:

```ocaml
(* build.mli *)
val existing_files : Path.Set.t -> ('a, bool) t
```

Where the boolean marks if all paths existed.

Reply:
Done, thanks for noticing! I only ported the `format_rules` occurrence, the one in build system looks too specific so I'll leave its generalization to another PR if you don't mind.